source/Tests/MessageContractAnalyzerFixture.cs

```diff
@@ -276,78 +276,100 @@ public class SimpleResponse : IResponse { }
     {
+        public SimpleRequest(string requiredString, int requiredInt, bool requiredBool, string[] requiredCollection)
+        {
+            RequiredStringProperty = requiredString;
+            RequiredIntProperty = requiredInt;
+            RequiredBoolProperty = requiredBool;
+            RequiredCollectionProperty = requiredCollection;
+
+            IntProperty = 27; // We don't decompile the constructor, it must be an inline initializer so this doesn't count
+        }
+
+        // --- sanity check required properties aren't affected -----
+
+        [Required]
+        public string RequiredStringProperty { get; set; }
+
+        [Required]
+        public int RequiredIntProperty { get; set; }
+
+        [Required]
+        public bool RequiredBoolProperty { get; set; }
+
+        [Required]
+        public string[] RequiredCollectionProperty { get; set; }
+
+        // --- important stuff starts here -----
+
         [Optional]
-        public string {|#0:StringProperty|} { get; set; }
+        public string {|#0:StringProperty|} { get; set; } // NOT OK, must be nullable or initialized
 
         [Optional]
-        public int {|#1:IntProperty|} { get; set; }
+        public string? NullableStringProperty { get; set; } // OK
 
         [Optional]
-        public string[] StringListProperty { get; set; } = null; // should NOT fire on this because optional nonnull collections are allowed
+        public string InitializedStringProperty { get; set; } =  ""foobar""; // OK
 
         [Optional]
-        public bool BoolProperty { get; set; } // should NOT fire on this because we have a special case for bool
-    }
-    /// <summary>response</summary>
-    public class SimpleResponse : IResponse { }
-}");
+        public string InitializedStringPropertyEmpty { get; set; } = string.Empty; // OK
 
-            await Verify.VerifyAnalyzerAsync(source,
-                new DiagnosticResult(Descriptors.OptionalPropertiesOnMessageTypesMustBeNullable).WithLocation(0).WithArguments("StringProperty", "string"),
-                new DiagnosticResult(Descriptors.OptionalPropertiesOnMessageTypesMustBeNullable).WithLocation(1).WithArguments("IntProperty", "int"));
-        }
+        [Optional]
+        public string InitializedStringPropertyExplicitNull { get; set; } = null!; // OK; If you really want to break things, then you can.
 
-        [Test]
-        public async Task MessageTypesMustInstantiateCollections()
-        {
-            var source = WithOctopusTypes(@"
-namespace Octopus.Core.Features.ServerTasks.MessageContracts
-{
-    /// <summary>request</summary>
-    public class SimpleRequest: IRequest<SimpleRequest, SimpleResponse> 
-    {
-        protected SimpleRequest() { }
-        public SimpleRequest(List<string> requiredStringList)
-        {
-            RequiredStringListProperty = requiredStringList;
-        }
+        [Optional]
+        public int {|#1:IntProperty|} { get; set; } // NOT OK, must be nullable or initialized.
```

Comment:
Should we have this twice - once where it's set in the ctor, and one where it's not?
Belt and braces and all that?

Reply:
🤷